### PR TITLE
Bump db version, "upgrade" path to new db in sled and indexeddb 

### LINKED ISSUE #756: Bumping version number of sled & indexeddb state store
https://github.com/matrix-org/matrix-rust-sdk/pull/729#discussion_r895594042_ changed the inner format of the sled state store, so did the changes in indexeddb. We need to bump the version number, ideally would provide some upgrade path...

### PR DESCRIPTION
**Pull request by gnunicorn**
Missing in #729: this bumps the db version and drops and recreates the effected trees/tables (for encrypted DBs) in sleds and indexeddbs StaetStore. Together with a fresh sync this should be enough of an upgrade path...

Fixes #756

### CODE REVIEW

**Review by jplatte — APPROVED** [review]
Looks reasonable but not at all my area of expertise, so @poljar if you have time maybe you should review too.

**Comment by poljar** [comment]
Looks fine, though has anybody tried this out? If not I would like to try how it behaves in a real client.

**Comment by zecakeh** [comment]
I just tested it in Fractal, and it panics at https://github.com/matrix-org/matrix-rust-sdk/blob/main/crates/matrix-sdk-base/src/rooms/normal.rs#L539.

I'm guessing it didn't expect to be called before the initial sync is done, because you shouldn't have any rooms locally before that.

If we're triggering an initial sync again, couldn't we erase the whole state store, since it's gonna be repopulated?

**Comment by gnunicorn** [comment]
> If we're triggering an initial sync again, couldn't we erase the whole state store, since it's gonna be repopulated?

I tried to avoid that. But if the rest of the outer infrastructure implies that, we might want to do that...

> I just tested it in Fractal, and it panics at https://github.com/matrix-org/matrix-rust-sdk/blob/main/crates/matrix-sdk-base/src/rooms/normal.rs#L539.

Well, that shouldn't panic in the first place. If there is no sync-token, it should error... `expect` in result-returning functions is really not a good pattern...

**Comment by zecakeh** [comment]
There would also be inconsistencies in the db with the current upgrade: `JOINED_USER_IDS` was reset, however in `ROOM_INFO`, `RoomInfo::members_synced` wasn't. Which means all the methods that use it are unreliable and it's not even possible to load missing members.

File: crates/matrix-sdk-sled/src/state_store.rs
Comment (zecakeh) [inline crates/matrix-sdk-sled/src/state_store.rs:227]:
```suggestion
            let path = path.join("matrix-sdk-state");
```

Although the name is inaccurate when it also contains the crypto store, it's the previous name used by `inner_open_with_path` in old L298. This would create a new database alongside the old one.

Another solution might be to rename the old database.

**Comment by zecakeh** [comment]
It would probably be useful to take the `MigrationConflictStrategy` as a parameter for `make_store_config`. I like that the default is non-destructive but if we don't have custom local-only data in the state store, there's not much use to keep a backup since we'll receive everything via sync again, so being able to chose the strategy without having to handle construction of the store ourselves seems useful.

File: crates/matrix-sdk-sled/src/lib.rs
Comment (zecakeh) [inline crates/matrix-sdk-sled/src/lib.rs:19]: We need to export `MigrationConflictStrategy` too.

File: crates/matrix-sdk-sled/src/state_store.rs
Comment (zecakeh) [inline crates/matrix-sdk-sled/src/state_store.rs:195]:
```suggestion
    SESSION,
    STRIPPED_INVITED_USER_ID,
```
Since it was changed from plain to encrypted, we also need to drop it.

**Review by zecakeh — CHANGES_REQUESTED** [review]
With those two changes, plus my previous suggestion, this PR works with Fractal.

File: labs/sled-state-inspector/src/main.rs
Comment (jplatte) [inline labs/sled-state-inspector/src/main.rs:7]: I would prefer if we didn't have to import builder types separately. Can we have `StateStore::builder()` instead / use it if we already have it?

File: crates/matrix-sdk-indexeddb/src/lib.rs
Comment (jplatte) [inline crates/matrix-sdk-indexeddb/src/lib.rs:33]:
```suggestion
    let mut builder = StateStore::builder();
```

File: crates/matrix-sdk-indexeddb/src/lib.rs
Comment (jplatte) [inline crates/matrix-sdk-indexeddb/src/lib.rs:65]:
```suggestion
        let mut builder = StateStore::builder();
```

File: crates/matrix-sdk-sled/Cargo.toml
Comment (jplatte) [inline crates/matrix-sdk-sled/Cargo.toml:48]: Should be moved up for consistent ordering, also extra space at the start.

File: crates/matrix-sdk-indexeddb/src/state_store.rs
Comment (jplatte) [inline crates/matrix-sdk-indexeddb/src/state_store.rs:272]: Should we really be using `derive_builder` for a four-parameter thing? I kind of feel like that just makes it much easier to accidentally have less docs, while hardly making things simpler. This builder doesn't have any useful docs on its methods as-is, right?

File: crates/matrix-sdk-sled/src/state_store.rs
Comment (jplatte) [inline crates/matrix-sdk-sled/src/state_store.rs:95]: Why?

File: crates/matrix-sdk-sled/src/state_store.rs
Comment (jplatte) [inline crates/matrix-sdk-sled/src/state_store.rs:234]: Can be `self.path.as_ref().map(...)`.

File: crates/matrix-sdk-sled/src/state_store.rs
Comment (jplatte) [inline crates/matrix-sdk-sled/src/state_store.rs:485]:
```suggestion
                // we stored some fields un-encrypted. Drop them to force re-creation
```

File: crates/matrix-sdk-test-macros/src/lib.rs
Comment (jplatte) [inline crates/matrix-sdk-test-macros/src/lib.rs:42]: Or just `.unwrap()`?

File: crates/matrix-sdk-indexeddb/src/state_store.rs
Comment (gnunicorn) [reply crates/matrix-sdk-indexeddb/src/state_store.rs:272]: [`derive_builder` has docs support](https://docs.rs/derive_builder/latest/derive_builder/#documentation-comments-and-attributes) and the compiler is complaining about them being missing.

grafik


Another reason I opted for using the derive_builder (other than we have that dependency already anyways), was to extending the configuration builder later a lot easier. Sure, it is only three params now, but it was only two before and through this we can very easily extend things without bloating our code a lot or breaking downstream compatibility.

File: crates/matrix-sdk-sled/src/state_store.rs
Comment (gnunicorn) [reply crates/matrix-sdk-sled/src/state_store.rs:234]: it doesn't as it also modified `cfg` and putting that into a `map` means it is moved, but then needed later. sure there is a way to do that, but it becomes icky...

File: crates/matrix-sdk-sled/src/state_store.rs
Comment (gnunicorn) [reply crates/matrix-sdk-sled/src/state_store.rs:248]: similarly here, we use `?` inside the `map` a few times, which either means we have to unwrap a result on the outside or handle it in another way. both are less legible IMHO.

File: crates/matrix-sdk-test-macros/src/lib.rs
Comment (gnunicorn) [reply crates/matrix-sdk-test-macros/src/lib.rs:42]: unwrap unfortunately often doesn't give us good enough information about the actual error. It only tells us that we tried to `unwrap` a result and then (b/c of the macro) links to the wrong code location. This assert on the other hand returns debuggable information of the actual cause of the error.

File: crates/matrix-sdk-sled/src/state_store.rs
Comment (jplatte) [reply crates/matrix-sdk-sled/src/state_store.rs:234]: No, closures are by-ref unless you use the `move` keyword. [Demo](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fd65b79903dfe3a1a0ac049b5282f590)

File: crates/matrix-sdk-indexeddb/src/lib.rs
Comment (gnunicorn) [reply crates/matrix-sdk-indexeddb/src/lib.rs:22]: why shouldn't we expose the type? you could use it directly, you don't have to go through `StateStore::builder`, why do you want to prevent that?

Comment (jplatte) [reply crates/matrix-sdk-indexeddb/src/lib.rs:22]: Oh, this is a `pub use`. Sorry.

File: crates/matrix-sdk-indexeddb/src/state_store.rs
Comment (jplatte) [reply crates/matrix-sdk-indexeddb/src/state_store.rs:272]: Alright, I guess it's not too bad.

File: crates/matrix-sdk/examples/autojoin.rs
Comment (jplatte) [inline crates/matrix-sdk/examples/autojoin.rs:52]:
```suggestion
        let state_store = matrix_sdk_sled::StateStore::builder().path(home).build()?;
```

same below and in the other example(s).

File: xtask/src/ci.rs
Comment (jplatte) [inline xtask/src/ci.rs:201]: I find this comment much less clear than the code it applies to. Maybe remove it?

**Review by jplatte — COMMENTED** [review]
GitHub is telling me I have two pending comments, either it's wrong or it threw away a bunch. I'll submit to see what happens, and continue reviewing after.

edit: It threw away a bunch of review comments (because I clicked 'Refresh') 🙁

File: crates/matrix-sdk-indexeddb/Cargo.toml
Comment (jplatte) [inline crates/matrix-sdk-indexeddb/Cargo.toml:28]: Let's please not do this. The futures crate isn't using the `crate?/feature` syntax (presumably because of MSRV), so activating the `alloc` feature is going to pull in basically all of the futures-* crates: https://github.com/rust-lang/futures-rs/blob/5c17408232e07fe6619ce7d501839ffdeab9c88f/futures/Cargo.toml#L38

File: crates/matrix-sdk-indexeddb/Cargo.toml
Comment (jplatte) [inline crates/matrix-sdk-indexeddb/Cargo.toml:40]: Please order alphabetically (I'll look into making this a CI check).

File: crates/matrix-sdk-indexeddb/src/state_store.rs
Comment (jplatte) [inline crates/matrix-sdk-indexeddb/src/state_store.rs:1679]:
```suggestion
        Ok(IndexeddbStore::builder().name(db_name).build().await?)
```

also below.

File: crates/matrix-sdk-sled/src/state_store.rs
Comment (jplatte) [inline crates/matrix-sdk-sled/src/state_store.rs:1731]:
```suggestion
        SledStore::builder().build().map_err(Into::into)
```

Comment (jplatte) [reply crates/matrix-sdk-sled/src/state_store.rs:1731]: Also a few times below.

File: crates/matrix-sdk-sled/src/state_store.rs
Comment (jplatte) [reply crates/matrix-sdk-sled/src/state_store.rs:236]: I don't insist.

Comment (gnunicorn) [reply crates/matrix-sdk-sled/src/state_store.rs:236]: hmmm... the demo doesn't seem applicable here... maybe because `string` is `copy`?.

When I do:
```rust
        let mut cfg = Config::new().temporary(is_temp);

        let path = self.path.as_ref().map(|origin_path| {
            let path = origin_path.join("matrix-sdk-state");

            cfg = cfg.path(&path);
            path
        });

        let db = cfg.open().map_err(StoreError::backend)?;

```

The compiler tells me:

```
warning: value assigned to `cfg` is never read
   --> crates/matrix-sdk-sled/src/state_store.rs:232:13
    |
232 |             cfg = cfg.path(&path);
    |             ^^^
    |
    = note: `#[warn(unused_assignments)]` on by default
    = help: maybe it is overwritten before being read?

error[E0382]: borrow of moved value: `cfg`
   --> crates/matrix-sdk-sled/src/state_store.rs:236:18
    |
227 |         let mut cfg = Config::new().temporary(is_temp);
    |             ------- move occurs because `cfg` has type `Config`, which does not implement the `Copy` trait
228 | 
229 |         let path =self.path.as_ref().map(|origin_path| {
    |                                          ------------- value moved into closure here
...
232 |             cfg = cfg.path(&path);
    |                   --- variable moved due to use in closure
...
236 |         let db = cfg.open().map_err(StoreError::backend)?;
    |                  ^^^^^^^^^^ value borrowed here after move

For more information about this error, try `rustc --explain E0382`.
warning: `matrix-sdk-sled` (lib) generated 1 warning
error: could not compile `matrix-sdk-sled` due to previous error; 1 warning emitted
```

Comment (gnunicorn) [reply crates/matrix-sdk-sled/src/state_store.rs:236]: (it's also what I tried before, when I made the previous comment)

File: crates/matrix-sdk-indexeddb/Cargo.toml
Comment (gnunicorn) [reply crates/matrix-sdk-indexeddb/Cargo.toml:40]: honestly not in favor of more ci-checks on minor style things (e.g. alphabetical order of dependencies in `Crates.toml`-files ... they are not that long...) that take more CI runs to fix and don't have an automated fixing mechanism that runs in seconds....

File: crates/matrix-sdk-sled/src/state_store.rs
Comment (jplatte) [reply crates/matrix-sdk-sled/src/state_store.rs:236]: No, `String` is not `Copy`. I'll try to understand what's going on here, but I don't want to block this PR on it.

**Review by jplatte — APPROVED** [review]

**Comment by gnunicorn** [comment]
I am locally seeing the same issue as we see on the CI with the wasm-tests. However, it works in non-headless mode and I don't really see any path forward for debugging it... trying to work around.

**Comment by gnunicorn** [comment]
looks like that is just a [badly phrased way to tell us it hit the timeout](https://github.com/rustwasm/wasm-bindgen/issues/1801) ... which can happen when we add more tests. bumping the timeout, hope that is it...

**Pull request merged**
